metal_header/device.c++

```diff
@@ -0,0 +1,140 @@
+/* Copyright 2018 SiFive, Inc */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#include "metal_header/device.h"
+
+#include <iostream>
+#include <set>
+#include <string>
+
+Device::Device(std::ostream &os, const fdt &dtb, std::string compat_string)
+  : os(os), dtb(dtb), compat_string(compat_string)
+{}
+
+void Device::emit_comment(const node &n) {
+  os << "/* From " << n.name() << " */\n";
+}
+
+void Device::emit_handle(std::string d, const node &n, std::string v) {
+  emit_comment(n);
+  os << "#define __METAL_DTB_HANDLE_" << d << "_" << n.handle() << " " << v << "\n";
+}
+
+void Device::emit_def_handle(std::string handle, const node &n, std::string field) {
+  emit_comment(n);
+  os << "#define " << handle << " (&__metal_dt_" << n.handle() << field << ")\n\n";
+  os << "#define " << "__METAL_DT_" << n.handle_cap() << "_HANDLE (&__metal_dt_"
+     << n.handle() << field << ")\n\n";
+}
+
+void Device::emit_def(std::string handle, std::string field) {
+  os << "#define " << handle << " " << field << "\n\n";
+}
+
+void Device::emit_def_value(std::string name, const node &n, std::string field) {
+  uint32_t irqs;
+  os << "#define __METAL_" << n.handle_cap() << "_INTERRUPTS \t\t";
+  if (name.compare("interrupts") == 0) {
+    irqs = n.get_fields_count<uint32_t>(name);
+  } else {
+    irqs = n.get_fields_count<std::tuple<node, uint32_t>>(name);
+  }
+  os << irqs << "\n";
+  os << "#define METAL_MAX_" << field << "\t __METAL_" << n.handle_cap() << "_INTERRUPTS\n\n";
```

Comment:
I think this method, should not emit the METAL_MAX_XXX but instead accumulated the interrupts, per field type.
So would need a switch statement for the field arg, and separate variable for each field type.

Then have a separate virtual method can emit the META_MAX_XXX with the accumulated value.